### PR TITLE
Add event to RuntimeEventSource for AppContext switches

### LINKED ISSUE #56142: [Feature Request]: Diagnostics for App Context switches
Original combined issue [here](https://github.com/dotnet/runtime/issues/56072)

The goal is to help customers better understand how their applications are running by adding events and counters that can give more insights and will help them debug potential performance issues. It'll also help us debug customer applications and write tools that would be able to glean insights from traces.

What app context switches are used, and what are they set to? This will tell customers what quirks are enabled (binary formatter etc.) This can be a huge aid when debugging applications.

### PR DESCRIPTION
**Pull request by agocke**
Fixes https://github.com/dotnet/runtime/issues/56142

### CODE REVIEW

**Comment by agocke** [comment]
@josalem Could you give me some advice on how to test this?

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/RuntimeEventSource.cs
Comment (josalem) [inline src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/RuntimeEventSource.cs:57]:
```suggestion
        [Event(1, Level = EventLevel.Informational)]
        internal void LogAppContextSwitch(string switchName, int value)
        {
            base.WriteEvent(1, switchName, value);
        }
```

It doesn't matter functionally, but I think this is the first event on this source, so it should get ID 1.

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/RuntimeEventSource.cs
Comment (josalem) [inline src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/RuntimeEventSource.cs:53]: For ease of use, it might be best to assign this a keyword. 

When you got to enable this, it'll create the counters. They won't get turned on unless you specify `EventCounterIntervalSec` as a filter parameter, though. That being said, we don't want to have the presence of the filter string be the switch for enabling counters or not.

You can do this by adding a nested class like this:
```csharp
public static class Keywords
{
	public const EventKeyword AppContext = (EventKeyword)0x1;
}
```

Then you can assign it to the event by updating the `EventAttribute` to have the `EventKeyword` parameter.

This way a user can turn on _just_ this event without having to turn on all the counters.

CC @brianrob to weigh in on the experience of this.

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/RuntimeEventSource.cs
Comment (josalem) [inline src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/RuntimeEventSource.cs:97]: The more traditional pattern is to store a static instance of the source

```csharp
public static RuntimeEventSource Log = new RuntimeEventSource();
```

and then access that anywhere you use it. So, the inverse of the access as written.

I think that means you'd need to change the visibility of the `s_SuntimeEventSource` field and potentially remove the `Initialize` method in exchange for the static initializer.

File: src/libraries/System.Private.CoreLib/src/System/AppContext.cs
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/System/AppContext.cs:147]: This is only going to log switches set using `AppContext.SetSwitch` . It won't log switches set using build properties and runtime.json that I think are much more common these days. Do we need to log those too?

**Review by josalem — COMMENTED** [review]
To test this, I would create an `EventListener` that enables this `EventSource` and then parses the `LogAppContextSwitch`.

Since you don't know what settings are going to be set in a given test leg, I'd use `AppContext.SetSwitch` to set a specific switch and look for it in the events. Alternatively, you could compare the events directly to the values in the dictionaries, but that might be harder to set up due to field visibility.

The listener might look like this:

```csharp
public sealed class AppContextListener : EventListener
{
	override protected void OnEventSourceCreated(EventSource source)
	{
		if (source.Name == "System.Runtime")
		{
			// the below will change if you add a keyword
			EnableEvents(source, EventLevel.Verbose, (EventKeywords)(-1));
		}
	}

	protected override void OnEventWritten(EventWrittenArgs eventData)
	{
		if (eventData.EventName == "LogAppContextSwitch" && eventData.Payload is not null)
		{
			string switchName = string.Empty;
			int switchValue = -1;
			for (int i = 0; i < eventData.Payload.Length; i++)
			{
				if (eventData.PayloadNames[i] == "switchName")
					switchName = eventData.Payload[i] as string;
				else if (eventData.PayloadNames[i] == "value")
					switchValue = eventData.Payload[i] as int;
			}

			if (switchName != string.Empty && switchValue != -1)
				// validate the switch event
		}
	}
}
```

File: src/libraries/System.Private.CoreLib/src/System/AppContext.cs
Comment (agocke) [reply src/libraries/System.Private.CoreLib/src/System/AppContext.cs:147]: Yes, I would like to include those as well -- do you know which API they go through if not SetSwitch?

Comment (jkotas) [reply src/libraries/System.Private.CoreLib/src/System/AppContext.cs:147]: `AppDomain.SetData` (https://github.com/dotnet/runtime/issues/47922 tracks adding `AppContext.SetData` for completeness)

Also, the runtime startup goes over everything that is in runtime.json and calls `AppDomain.SetData` for it.

Comment (agocke) [reply src/libraries/System.Private.CoreLib/src/System/AppContext.cs:147]: In that case I will also add everything in `s_dataStore` which has a boolean value. I'm trying to avoid grabbing everything in s_dataStore to limit the amount of potential user data that's recorded.

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/RuntimeEventSource.cs
Comment (josalem) [reply src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/RuntimeEventSource.cs:97]: I took a look at where it's initialized (`System.StartupHookProvider.ProcessStartupHooks`). I don't think there would be any side-effects related to removing the `Initialize` method and relying on the static initializer. I'm okay leaving it as is, though.

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/RuntimeEventSource.cs
Comment (agocke) [reply src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/RuntimeEventSource.cs:53]: Sorry, I'm not understanding this at all. Are you saying that we should add an additional registration key for just grabbing the AppContext switches?

Comment (brianrob) [reply src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/RuntimeEventSource.cs]: Let me try to re-state @josalem's position and see if I get it right.  I think that @josalem is saying that right now, everything in this EventSource gets turned on all at once because there is no ability to filter.  The counters get created but not actually enabled because a filter string is required to turn them on, but I think he's saying that in the future, he'd like for this not to be the case, and so wants to future proof this with a keyword.  Keywords are one of the constructs that let you filter, and so I believe he's just proposing adding a keyword to allow this to be turned on and off independently of the counters. @josalem did I get that right?

I have another question on this - is this where we would want this and other future managed code events to live, when their functionality is in CoreLib, or do we think that the event should live somewhere like `FrameworkEventSource`?

Comment (brianrob) [reply src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/RuntimeEventSource.cs]: I should also add that I think this event is a great idea, and I will absolutely turn this on by default in PerfView.

Comment (agocke) [reply src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/RuntimeEventSource.cs]: I think this is *a* reasonable place for the AppContext switch events to live -- I'm not sure I'd be bold enough to take a stronger position than that.

**Comment by agocke** [comment]
Is there an easy way to run just the tracing tests (src/tests/tracing/runtimeeventsource/*) locally? It looks like there's an existing event listener in there that I'd like to augment, but I'm having trouble running just those test cases

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/RuntimeEventSource.cs
Comment (brianrob) [reply src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/RuntimeEventSource.cs]: Fair enough.  I was aiming that question more at @josalem.  :)

Comment (josalem) [reply src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/RuntimeEventSource.cs]: Your assessment is spot-on, Brian.

I think it is fine for now. If we add a keyword for the AppContext switches, we can at least turn on just those events. Not specifying `EventCounterIntervalSec` means we can turn on the world without enabling the counters, but it's not ideal as an enabling mechanism.

**Comment by josalem** [comment]
> Is there an easy way to run just the tracing tests (src/tests/tracing/runtimeeventsource/*) locally? 

The fastest way I've used for running the tests is to build them directly and bypass the xunit wrapper cruft:
1. `build.{cmd|sh} clr+libs -rc checked -lc release`
2. `./src/tests/build.{cmd|sh} generatelayoutonly checked`
3. `pushd ./src/tests/tracing/runtimeeventsource`
4. `../../../.dotnet/dotnet build <test to build>.csproj -c Checked`
5. `pushd <path to built test>`
6. set any env vars you need (I think this test only needs `COMPlus_TieredCompilation=0`)
7. `runtimeeventsource.{cmd|sh} -coreroot=<path from step 2>`

It's ugly, but better than building _all_ the tests. If anyone knows a better way...

**Comment by stephentoub** [comment]
Various AppContext switches represent something that's also settable via environment variables.  Do we have the ability to reflect those as well easily?  I understand we likely don't want to just dump out all environment variables, to avoid dumping out arbitrary user data.

**Comment by agocke** [comment]
> ../../../.dotnet/dotnet build <test to build>.csproj -c Checked

No dice. This seems to just restore the project -- it doesn't build anything

**Comment by josalem** [comment]
>  it doesn't build anything

It didn't drop the built test somewhere in the artifacts directory?

**Comment by agocke** [comment]
Nope. From MSBuild it looks like it imported `tests/Common/nobuild.targets` so it must think it didn't need to build. I'll see if I can figure out why

**Comment by agocke** [comment]
Still no idea why it didn't build, but passing `-p:_WillCLRTestProjectBuild=true` overrode the logic

File: src/libraries/System.Private.CoreLib/src/System/AppContext.cs
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/System/AppContext.cs:163]: This should use the same logic as `TryGetSwitch` - look for `if (GetData(switchName) is string value && bool.TryParse(value, out isEnabled))`

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/RuntimeEventSource.cs
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/RuntimeEventSource.cs:61]: Insight into GitHub's C# colorizer algorithm:
image
:smile:

File: src/libraries/System.Private.CoreLib/src/System/AppContext.cs
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/System/AppContext.cs:138]: This ifdef should not be needed. AppContext.cs is only ever compiled into CoreLib that has `NETCOREAPP` defined.

Comment (agocke) [reply src/libraries/System.Private.CoreLib/src/System/AppContext.cs:138]: I was trying to solve the problem with browser, where it seems `RuntimeEventSource` doesn't exist -- but maybe this isn't the right ifdef...

**Comment by agocke** [comment]
I think this is ready for re-review

**Comment by agocke** [comment]
ping @josalem @brianrob for review

**Comment by josalem** [comment]
Looking now

File: src/libraries/System.Private.CoreLib/src/System/AppContext.AnyOS.cs
Comment (josalem) [inline src/libraries/System.Private.CoreLib/src/System/AppContext.AnyOS.cs:62]: The event doesn't seem to differentiate between switches that come from `s_dataStore` and `s_switches`, but I believe the lookup logic only falls back to `s_dataStore` if a requested key _doesn't_ exist in `s_switches`.

We should either differentiate between the two stores as part of the event or only send events for `s_dataStore` if a given key _doesn't_ exist in `s_swtiches`. Otherwise, we can see duplicate keys and have no way of knowing which took precedence.

File: src/tests/tracing/eventcounter/runtimecounters.cs
Comment (josalem) [inline src/tests/tracing/eventcounter/runtimecounters.cs:47]: You can prevent these from showing up by excluding the keyword you added from the call to `EnableEvents` down on L61.  Something like

```csharp
EnableEvents(source, EventLevel.Informational, (EventKeywords)((-1) & ~(int)(RuntimeEventSource.Keywords.AppContext)), refreshInterval);
```

File: src/tests/tracing/eventcounter/runtimecounters.cs
Comment (josalem) [inline src/tests/tracing/eventcounter/runtimecounters.cs:124]: If possible, I'd want these in a separate test since a test failure here will get incorrectly bucketed as an `EventCounters` failure. If that's going to take too long, we can leave it here though.

**Review by josalem — APPROVED** [review]
Another round of comments. I'm going to approve it assuming you address the comment about switch precedence, so we can expedite getting this merged 😄

File: src/tests/tracing/eventcounter/runtimecounters.cs
Comment (agocke) [reply src/tests/tracing/eventcounter/runtimecounters.cs:47]: Is this something I would need only if we split this into a different test? Right now I want to see all events

File: src/tests/tracing/eventcounter/runtimecounters.cs
Comment (agocke) [reply src/tests/tracing/eventcounter/runtimecounters.cs:124]: This doesn't look like it will be ready for the 12pm snap, so I can refactor.

Comment (josalem) [reply src/tests/tracing/eventcounter/runtimecounters.cs:124]: I'm okay leaving it as-is if you want to merge it now.

Comment (agocke) [reply src/tests/tracing/eventcounter/runtimecounters.cs:124]: Tests weren't complete by that point, so I think it's a moot point.

I'll move things over to a new class

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/RuntimeEventSource.cs
Comment (davidfowl) [inline src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/RuntimeEventSource.cs:64]: @brianrob Would a single event with all of the data be better for something like this?

Comment (brianrob) [reply src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/RuntimeEventSource.cs:64]: I think one event per switch is good.  It makes it easy to consume the data, and I can't imagine we'll have so many that the cost of this event will be painful.  If/when we choose to create events were we dump data in bulk, it makes both the event production and event consumption code more complex.  On the production side, you have to chunk the data to ensure that you don't run up against size limitations (e.g. ETW has a 64K size limit per event).  On the consumption side, you have to be more sophisticated in how you parse the event.  There are definitely places were it's valuable - for example BulkType events in the runtime when we log heap snapshots.  There are just tons of types, and so emitting each one in its own event is very costly.

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/RuntimeEventSource.cs
Comment (elinor-fung) [inline src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/RuntimeEventSource.cs:71]: Does this enable event command still come through with startup tracing?
(for example `dotnet trace collect --providers System.Runtime:1 -- app.exe`)

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/RuntimeEventSource.cs
Comment (elinor-fung) [inline src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/RuntimeEventSource.cs:64]: Is having the `Log` prefix in the event name normal? I think I have more commonly seen names like just `AppContextSwitch` or `AppContextSwitchInfo`.

File: src/tests/tracing/runtimeeventsource/RuntimeEventSourceTest.cs
Comment (agocke) [reply src/tests/tracing/runtimeeventsource/RuntimeEventSourceTest.cs:24]: Would you increase the timeout or do something else? In theory, when we create the runtime listener it should immediately start dumping the AppContext events, so I don't think the delay should even be necessary. But I don't know the internals of EventSource.

Comment (elinor-fung) [reply src/tests/tracing/runtimeeventsource/RuntimeEventSourceTest.cs:24]: Ah, I think you may be right about the delay not being necessary, since the AppContext events are dumped synchronously on event source enable. (@josalem?)

My mind was on other tracing scenarios that are capturing events for some action not the current state at enable. In those cases, I've seen (and written) tests spin waiting until we get some expected event or events.

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/RuntimeEventSource.cs
Comment (elinor-fung) [inline src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/RuntimeEventSource.cs:107]: This is just logging values at a specific point in time (when the event source is enabled) - is that the desire?

I'm thinking of the possibility of scenarios like:
- Initial values do not have `SwitchX` set
- Event source is enabled, logging that `SwitchX` is not set
- App sets `SwitchX`
- Runtime (or app/library) reads `SwitchX` and is now configured a certain way

or

- Initial values do not have `SwitchX` set
- Runtime (or app/library) reads `SwitchX` and is now configured a certain way
- App sets `SwitchX`
- Event source is enabled, logging that `SwitchX` is set

Comment (davidfowl) [reply src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/RuntimeEventSource.cs:107]: I think most of those are edge cases. The 90% usage is:
- Switch set in the entry point
- Switch set in runtime.config.json (which comes from the csproj)

File: src/tests/tracing/runtimeeventsource/RuntimeEventSourceTest.cs
Comment (brianrob) [reply src/tests/tracing/runtimeeventsource/RuntimeEventSourceTest.cs:24]: Agreed, you should not need the `Task.Delay` call.  All writes to an `EventListener` are synchronous on the thread that does the write.

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/RuntimeEventSource.cs
Comment (elinor-fung) [reply src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/RuntimeEventSource.cs:107]: Definitely agree that those should not be the common case, but wanted to call it out.

> - Switch set in the entry point

I think part of my question is what does this end up looking like when tracing startup (where I'd expect the event source is enabled before the app's entry is is run)? Could it end up being the first scenario I had above?

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/RuntimeEventSource.cs
Comment (agocke) [reply src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/RuntimeEventSource.cs:71]: Good question, but out of my wheelhouse and I don't really know how to test it. @brianrob?

Comment (brianrob) [reply src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/RuntimeEventSource.cs:71]: `OnEventCommand` should be fired whenever an `EventSource` is enabled, and exists in the process.  So on start-up, when the `EventSource` comes into existence if `dotnet trace` has already started collection, then the `EventSource` will be registered with EventPipe, be told that it's enabled, and then call this code.  So the short answer (after the long answer) is yes.

Comment (elinor-fung) [reply src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/RuntimeEventSource.cs:71]: > how to test it

I'd think you could create an app that has some feature switch set in its runtimeconfig.json, set the `DOTNET_ROOT` environment variable to point to your local build, and then collect a trace with `dotnet trace collect --providers System.Runtime:1 -- app.exe`

File: src/libraries/System.Private.CoreLib/src/System/AppContext.AnyOS.cs
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/AppContext.AnyOS.cs:46]: Nit:
```suggestion
                    foreach (KeyValuePair<string, bool> kvp in s_switches)
                    {
                        // Convert bool to int because it's cheaper to log (no boxing)
                        ev.LogAppContextSwitch(kvp.Key, kvp.Value ? 1 : 0);
```

File: src/libraries/System.Private.CoreLib/src/System/AppContext.AnyOS.cs
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/AppContext.AnyOS.cs:67]:
```suggestion
                    static void LogDataStore(Dictionary<string, bool>? switches)
```

File: src/libraries/System.Private.CoreLib/src/System/AppContext.AnyOS.cs
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/AppContext.AnyOS.cs:75]: Nit:
```suggestion
                        foreach (KeyValuePair<string, object?> kvp in s_dataStore)
                        {
                            if (kvp.Value is string s &&
                                bool.TryParse(s, out bool isEnabled) &&
                                switches?.ContainsKey(kvp.Key) != true)
                            {
                                ev.LogAppContextSwitch(kvp.Key, isEnabled ? 1 : 0);
```

**Comment by davidfowl** [comment]
@agocke Can you make sure this goes into the 6.x branches as well once we merge?

File: src/libraries/System.Private.CoreLib/src/System/AppContext.AnyOS.cs
Comment (agocke) [reply src/libraries/System.Private.CoreLib/src/System/AppContext.AnyOS.cs:67]: Captures s_dataStore and `ev` parameter

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/RuntimeEventSource.cs
Comment (agocke) [reply src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/RuntimeEventSource.cs:64]: It's internal so I didn't think too hard. We can certainly change it if it becomes annoying later.

Comment (elinor-fung) [reply src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/RuntimeEventSource.cs:64]: I was thinking it was external in that it is the name of the event that gets fired.

File: src/tests/tracing/runtimeeventsource/RuntimeEventSourceTest.cs
Comment (elinor-fung) [inline src/tests/tracing/runtimeeventsource/RuntimeEventSourceTest.cs:80]: Maybe have separate checks and failure messages not having switch logged at all versus having the switch logged with the wrong value?

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/RuntimeEventSource.cs
Comment (agocke) [reply src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/RuntimeEventSource.cs:71]: Looks like it works -- I got the AppContext switch out when I set it via `RuntimeHostConfigurationOption`

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/RuntimeEventSource.cs
Comment (davidfowl) [reply src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/RuntimeEventSource.cs:107]: I'm not sure, @brianrob @josalem do event sources OnCommand run before Main executes when you attach the event source at startup (which not many do). 

@elinor-fung These are great questions BTW 👍🏾

Comment (brianrob) [reply src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/RuntimeEventSource.cs:107]: They definitely can, but not all will.  If the EventSource comes into existence before Main is called and it is told to turn on (e.g. EventPipe, ETW, or some EventListener tells it to), then it will trigger before Main.  By design, there isn't anything that requires that Main run before events can be emitted.  A good example here is FrameworkEventSource which often comes into existence before Main.

Comment (brianrob) [reply src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/RuntimeEventSource.cs:107]: To add to this, if you want to ensure that these can be emitted early, make sure that the EventSource that emits them comes into existence early.  That may already be the case, but if not, make sure it gets instantiated early, even if it doesn't log any events when instantiated.  Just the instantiation will allow it to be enabled immediately (at least by ETW, but shortly by EventPipe once it comes up).

Comment (brianrob) [reply src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/RuntimeEventSource.cs:107]: @elinor-fung, just read your first scenario - I think that the right way to handle a race between logging the current state and changes to switches is to dump the set of switches at the beginning of the tracing session as this code does.  Additionally, any time a switch is configured, it should be logged immediately.  This ensures that if switches change value or get set after the session is created, we see the change.  This, combined with instantiating the EventSource early will ensure that we don't miss any data.

File: src/libraries/System.Private.CoreLib/src/System/AppContext.AnyOS.cs
Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/System/AppContext.AnyOS.cs:67]: s_dataStore is static; there's no conflict with the method being static and using s_dataStore. But I missed that it captured ev; if it were me, I'd just add `, RuntimeEventSource ev` as an argument and pass it in rather than capturing anything.

File: src/tests/tracing/runtimeeventsource/RuntimeEventSourceTest.cs
Comment (agocke) [reply src/tests/tracing/runtimeeventsource/RuntimeEventSourceTest.cs:80]: Not opposed, but I'm hoping to take this for RC 1, so I'd prefer to make test improvements in a later iteration

**Comment by agocke** [comment]
@josalem @brianrob  I think this is ready, could you take a last look?

**Review by brianrob — APPROVED** [review]
LGTM.  Also, confirmed that `RuntimeEventSource` is initialized in `ProcessStartupHooks`, so you should always get switches logged if requested prior to `Main`.

**Comment by lewing** [comment]
@agocke this appears to have broken:

runtime (Build Android x64 Release AllSubsets_Mono_RuntimeTests)
runtime (Build Browser wasm Release AllSubsets_Mono_RuntimeTests)

**Comment by agocke** [comment]
Woops, this shouldn't be running on those platforms. I thought I had successfully disabled it but apparently not. I'll fix.

**Pull request merged**
